File: R_Programming/ProgrammingAssignment3/rankhospital.R
rankhospital <- function(state, outcome, num='best') {
    # heart failure - column 17
    # heart attack  - column 11
    # pneumonia     - column 23
    list_outcome  <- list("heart failure" = 17, "heart attack" = 11, "pneumonia" = 23)
    outcome_file  <- 'Hospital/outcome-of-care-measures.csv'
    hospital_file <- 'Hospital/hospital-data.csv'
    
    stats <- read.csv(outcome_file, colClasses='character')  
    states <- levels(factor(stats[,7]))
    columns <- names(stats)
    
    if (!state %in% states) stop("invalid state")
    loc_outcome <- list_outcome[outcome]
    if (length(loc_outcome) == 0) stop("invalid outcome")
    loc_outcome <- as.numeric(loc_outcome)
    
    # split the entries that only belong to the state
    sub_stats <- subset(stats, State == state, select=c(columns[2], columns[loc_outcome]))
    sub_stats[, 2] <- suppressWarnings(as.numeric(sub_stats[, 2]))
    sub_stats <- na.omit(sub_stats)
    
    # num is 'best' or 'worst' or numeric
    list_rank <- list('best' = 1, 'worst' = nrow(sub_stats))
    if (is.character(num)) rank_out <- list_rank[[num]]
    else rank_out <- num
    
    # if rank_out is greater than the number of entries, return NA
    # otherwise reorder the data frame
    if (rank_out > nrow(sub_stats)) NA
    else {
        sub_stats <- data.frame(sub_stats[ order(sub_stats[,2], sub_stats[,1]), ], row.names=seq(nrow(sub_stats)))
        sub_stats[rank_out, 'Hospital.Name']    
    }
}

File: R_Programming/ProgrammingAssignment3/rankhospital.py
import pandas as pd
import numpy as np
from os import path, listdir


def rankhospital(state, outcome, num='best'):
    """
    """

    # heart failure - column 17
    # heart attack  - column 11
    # pneumonia     - column 23
    
    outcome_file = "Hospital/outcome-of-care-measures.csv"
    if not path.exists(outcome_file):
        raise IOError
    stats = pd.read_csv(outcome_file)
    dict_outcome = {'heart failure': 16, 'heart attack': 10, 'pneumonia': 22}
    dict_rank = {'best' : 1}
    columns = stats.columns.values

    allStates = stats.State.unique().astype(str).tolist()
    if not state in allStates:
        raise Exception('invalid state')
    if not outcome in dict_outcome.keys():
        raise Exception('invalid outcome')

    loc_outcome = dict_outcome[outcome]
    stats.iloc[:, loc_outcome] = stats.iloc[:, loc_outcome].convert_objects(convert_numeric=True)

    stats1 = stats.dropna(axis=0, subset=[ columns[loc_outcome] ]).sort([columns[loc_outcome], columns[1]], ascending=True)

    def ranker(df):
        df['state_rank'] = np.arange(len(df)) + 1
        return df

    stats2 = stats1.groupby('State').apply(ranker)
    stats3 = stats2[stats2.State == state]

    dict_rank.update({'worst' : len(stats3)})
    if isinstance(num, str):
        rank_out = dict_rank[num]
    else:
        rank_out = num
    if rank_out > len(stats3):
        print "NA"
        return

    ret = stats3[stats3.state_rank==rank_out].iloc[:, 1]
    print ret

    
    
    
        
    
